third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -649,6 +649,11 @@ impl<'a> FunctionGenerator<'a> {
 
     /// Generate code for the load instruction.
     fn gen_load(&mut self, ctx: &BytecodeContext, dest: &TempIndex, cons: &Constant) {
+        if let Some(pos) = self.stack.iter().position(|e| e == dest) {
```

Comment:
This file could really use some documented invariants, which would make it easier to reason logically about the code without deep understanding.  Using pattern-matching instead, I note a discrepancy between `gen_load` and other instruction implementations that makes me suspect that this is not the right solution.

Notably, `gen_call()` ends with a call 
```
        self.abstract_push_result(ctx, dest);`
```
while `gen_load()` ends with
```
        self.abstract_push_result(ctx, vec![*dest]);
```
and I don't see why these are not identical.  What's the difference?

Reply:
> This file could really use some documented invariants, which would make it easier to reason logically about the code without deep understanding.

Agree, but I think it is out of scope for this PR for me to add invariants throughout the file.

> I don't see why these are not identical. What's the difference?

Look at the types of `dest` in those two usages: one is a `&TempIndex`, the other is a `&[TempIndex]`.

Reply:
I mentioned already a few times, as early as end of last year, that this module needs to be extensively rewritten, if not replaced by a different design. (E.g. generating naive stack code, then running a per-block level optimization on the stack  bytecode). Lets not try to improve fundamentals of deprecated code.

There are two problems with this code:

- complexity and henceforth buginess
- very mediocre generated stack code, we have multiple bugs open with examples.